<a href="https://colab.research.google.com/github/ottrindade1963/analise_desenvolvimento/blob/main/Pipeline_AFRIMIDLE_MO3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pipeline Africa & Middle East v4.0
Walk-Forward CV | Optuna | MLflow | Ablação | GitHub + Colab

**Versão corrigida (enxuta) — 21 de julho de 2026.** Esta versão do notebook
**pressupõe que o repositório GitHub já foi actualizado** com os 6 ficheiros
corrigidos entregues em `africa_mo_v4_codigo_corrigido.zip` (relatório de
diagnóstico de causas-raiz + relatório de verificação das correções). Ela não
reescreve o código a cada execução — apenas clona/actualiza o repositório e
confirma que as correções estão presentes.

Se preferir uma versão que aplica as correções sozinha, sem depender de você
ter actualizado o GitHub manualmente, use a versão completa
(`Pipeline_AFRIMIDLE_MO_corrigido.ipynb`) em vez desta.

**⚠️ Aviso de segurança:** o notebook original continha um Personal Access Token
do GitHub escrito directamente no código (`ghp_...`). Esse token foi removido
desta versão — em vez disso, o notebook agora pede o token de forma segura
(via `getpass`, que não fica visível nem gravado no ficheiro `.ipynb`).
**Recomenda-se fortemente revogar o token antigo em
https://github.com/settings/tokens e gerar um novo**, já que o token anterior
ficou visível em texto simples no ficheiro que foi partilhado.

**Antes de começar:**
1. Actualize o repositório GitHub com os 6 ficheiros do zip de código corrigido
   (`utils/model_io.py`, `features/engineer.py`, `validation/walk_forward.py`,
   `explainability/ablation.py`, `explainability/innovations.py`, `pipeline.py`)
2. Runtime → Change runtime type → GPU (T4)
3. Executar as células por ordem — a Célula 2 vai pedir o seu utilizador e token do GitHub
4. A Célula 2.1 confirma que as correções chegaram ao Colab — não a saltar


## Célula 1 — Montar Google Drive

In [1]:
from google.colab import drive
drive.mount('/content/drive')
print("Drive montado.")


Mounted at /content/drive
Drive montado.


## Célula 2 — Clonar GitHub (CORRIGIDA — diagnóstico rápido)

**Se antes esta célula parecia "ficar à espera" depois de introduzir o token:**
isso normalmente acontece porque o `git` (chamado via `os.system`, sem
verificação de erro) tenta abrir um prompt interactivo de utilizador/senha
quando a autenticação embutida na URL falha — e o Colab não tem como responder
a esse prompt, por isso a célula fica presa indefinidamente, sem imprimir
nenhum erro.

Esta versão corrige isso de duas formas: (1) valida o token contra a API do
GitHub *antes* de tentar clonar, falhando rápido com uma mensagem clara se o
token estiver errado, expirado ou sem o scope `repo`; (2) define
`GIT_TERMINAL_PROMPT=0`, que impede o `git` de abrir esse prompt escondido —
se a autenticação falhar, ele erra imediatamente em vez de esperar para
sempre.

In [2]:
import os, sys, subprocess, json
from getpass import getpass
import urllib.request

# CORRECTION (segurança): o token deixou de estar escrito no código.
# É pedido de forma segura a cada execução — não fica gravado no notebook.
GITHUB_USER  = input("Utilizador do GitHub: ").strip() or "ottrindade1963"
GITHUB_TOKEN = getpass("Personal Access Token do GitHub (não fica visível): ").strip()
REPO_NAME    = "analise_desenvolvimento"
LOCAL_DIR    = "/content/africa_mo_v4"

# 1) Validar o token ANTES de tentar clonar — falha em segundos, com uma
# mensagem clara, em vez de o git ficar preso à espera de uma resposta que o
# Colab não consegue dar.
print("A validar o token junto da API do GitHub...")
req = urllib.request.Request(
    "https://api.github.com/user",
    headers={"Authorization": f"token {GITHUB_TOKEN}"}
)
try:
    with urllib.request.urlopen(req, timeout=15) as resp:
        user_info = json.loads(resp.read())
    print(f"✓ Token válido — autenticado como: {user_info.get('login')}")
except Exception as e:
    raise SystemExit(
        f"✗ Token inválido, expirado ou sem permissão ({e}).\n"
        "Pare aqui: gere um novo token em https://github.com/settings/tokens "
        "com o scope 'repo' marcado, e corra esta célula outra vez."
    )

# 2) Clonar. GIT_TERMINAL_PROMPT=0 impede o git de abrir um prompt
# interactivo que o Colab não consegue responder — se a autenticação ou o
# nome do repositório estiverem errados, ele erra imediatamente com uma
# mensagem, em vez de "ficar à espera" silenciosamente.
env = dict(os.environ, GIT_TERMINAL_PROMPT="0")
clone_url = f"https://{GITHUB_USER}:{GITHUB_TOKEN}@github.com/{GITHUB_USER}/{REPO_NAME}.git"

print(f"\nA clonar/actualizar '{REPO_NAME}'...")
if not os.path.exists(LOCAL_DIR):
    result = subprocess.run(["git", "clone", clone_url, LOCAL_DIR],
                             env=env, capture_output=True, text=True, timeout=120)
else:
    result = subprocess.run(["git", "pull"], cwd=LOCAL_DIR,
                             env=env, capture_output=True, text=True, timeout=120)

# Nunca mostrar o token em claro, mesmo dentro de uma mensagem de erro do git
safe_out = result.stdout.replace(GITHUB_TOKEN, "***")
safe_err = result.stderr.replace(GITHUB_TOKEN, "***")

if result.returncode != 0:
    print("✗ git falhou:")
    print(safe_out)
    print(safe_err)
    raise SystemExit(
        "Verifique: (1) REPO_NAME e o utilizador estão correctos "
        f"(REPO_NAME='{REPO_NAME}', utilizador='{GITHUB_USER}')? "
        "(2) o token tem o scope 'repo'? (3) o repositório existe e não "
        "pertence a outra conta/organização?"
    )
print("✓ Repositório clonado/actualizado com sucesso.")

# Corrigir estrutura, caso o clone tenha criado uma sub-pasta duplicada
sub = os.path.join(LOCAL_DIR, "africa_mo_v4")
if os.path.exists(sub):
    import shutil
    for item in os.listdir(sub):
        shutil.move(os.path.join(sub, item), os.path.join(LOCAL_DIR, item))
    os.rmdir(sub)
    print("Estrutura corrigida (sub-pasta duplicada removida).")

os.chdir(LOCAL_DIR)
if LOCAL_DIR not in sys.path:
    sys.path.insert(0, LOCAL_DIR)

del GITHUB_TOKEN   # não manter o token em memória mais do que o necessário


Utilizador do GitHub: ottrindade1963
Personal Access Token do GitHub (não fica visível): ··········
A validar o token junto da API do GitHub...
✓ Token válido — autenticado como: ottrindade1963

A clonar/actualizar 'analise_desenvolvimento'...
✓ Repositório clonado/actualizado com sucesso.
Estrutura corrigida (sub-pasta duplicada removida).


## Célula 2.0.1 — Instalar dependências

Separada da célula de clone para que fique claro onde está o tempo de espera:
esta instalação normalmente demora **3 a 10 minutos** num runtime novo do
Colab, porque `mlflow`, `pymc`, `xgboost`, `statsmodels`, `shap` e `arviz` são
pacotes pesados com muitas dependências — isto é normal e não significa que
está encravado. Se passar de uns 20 minutos sem terminar, interrompa a célula
(botão ■) e corra os pacotes em pequenos grupos para identificar qual está a
travar.

In [3]:
import subprocess, time

print("A instalar dependências (normalmente 3–10 min num runtime novo)...")
t0 = time.time()
subprocess.run(
    ["pip", "install", "-q",
     "wbgapi", "optuna", "shap", "pmdarima",
     "pymc", "arviz", "ruptures", "mlflow", "xgboost", "statsmodels"],
    check=True
)
print(f"✓ Dependências instaladas em {time.time() - t0:.0f}s.")


A instalar dependências (normalmente 3–10 min num runtime novo)...
✓ Dependências instaladas em 24s.


## Célula 2.1 — Confirmar que o repositório clonado tem as correções

Ao contrário da versão completa do notebook, esta célula **não escreve
nenhum ficheiro** — ela só lê o que foi clonado do GitHub e verifica se os 6
ficheiros corrigidos estão de facto lá. Se algum item aparecer com ✗, o
repositório ainda não foi actualizado com o zip de código corrigido — volte
a subir os ficheiros ao GitHub e rode a Célula 2 outra vez (`git pull`).

In [4]:
import inspect, importlib, sys

for mod in list(sys.modules.keys()):
    if any(x in mod for x in ["config", "validation", "tuning", "models",
                               "features", "explainability", "utils", "pipeline"]):
        del sys.modules[mod]

checks_ok = True
try:
    import utils.model_io as model_io
    has_bundle_fns = hasattr(model_io, "save_model_bundle") and hasattr(model_io, "load_model_bundle")
    print(f"  {'✓' if has_bundle_fns else '✗'} utils/model_io.py  →  save_model_bundle()/load_model_bundle()")
    checks_ok &= has_bundle_fns
except ImportError as e:
    print(f"  ✗ utils/model_io.py não encontrado: {e}")
    checks_ok = False

import features.engineer as engineer_mod
src = inspect.getsource(engineer_mod)
ok = "wgicomp" in src and "CORRECTION" in src
print(f"  {'✓' if ok else '✗'} features/engineer.py  →  fallback 'wgicomp' para A5 presente")
checks_ok &= ok

import validation.walk_forward as wf_mod
src = inspect.getsource(wf_mod)
ok = "save_model_bundle" in src
print(f"  {'✓' if ok else '✗'} validation/walk_forward.py  →  grava bundle (não modelo isolado)")
checks_ok &= ok

import explainability.ablation as ablation_mod
src = inspect.getsource(ablation_mod)
ok = "load_model_bundle" in src and "A1_WDI_only" in src
print(f"  {'✓' if ok else '✗'} explainability/ablation.py  →  lê bundle + baseline explícita")
checks_ok &= ok

import explainability.innovations as innovations_mod
src = inspect.getsource(innovations_mod)
ok = "scaler.transform" in src
print(f"  {'✓' if ok else '✗'} explainability/innovations.py  →  escalona antes do predict()")
checks_ok &= ok

import pipeline as pipeline_mod
src = inspect.getsource(pipeline_mod)
ok = "REFERENCE_SPEC_FOR_KERNEL_SHAP" in src
print(f"  {'✓' if ok else '✗'} pipeline.py  →  nome de spec actualizado + RMSE rotulado")
checks_ok &= ok

print()
if checks_ok:
    print("✓ Todas as correções estão presentes no repositório clonado.")
else:
    print("✗ Pelo menos uma correção está em falta — actualize o GitHub com o "
          "zip de código corrigido antes de continuar (ou use a versão completa "
          "do notebook, que aplica as correções automaticamente).")


  ✓ utils/model_io.py  →  save_model_bundle()/load_model_bundle()
  ✓ features/engineer.py  →  fallback 'wgicomp' para A5 presente
  ✓ validation/walk_forward.py  →  grava bundle (não modelo isolado)
  ✓ explainability/ablation.py  →  lê bundle + baseline explícita
  ✓ explainability/innovations.py  →  escalona antes do predict()
  ✓ pipeline.py  →  nome de spec actualizado + RMSE rotulado

✓ Todas as correções estão presentes no repositório clonado.


## Célula 3 — Verificar configuração

In [5]:
import os, sys
LOCAL_DIR = "/content/africa_mo_v4"
if LOCAL_DIR not in sys.path:
    sys.path.insert(0, LOCAL_DIR)
os.chdir(LOCAL_DIR)

import config.paths     as paths
import config.variables as var
import config.features  as feat
import config.model_params as mp

print(f"ROOT: {paths.ROOT}")
print(f"Target: {var.TARGET}")
print(f"Countries: {len(var.COUNTRY_CODES)}")
print(f"Specs: {list(feat.ABLATION_SPECS.keys())}")
print(f"LSTM lookback: {mp.LSTM['lookback']} anos")
print(f"Optuna trials: {mp.RF['n_trials']}")


ROOT: /content/africa_mo_v4
Target: valor_agregado_industrial_percent_pib
Countries: 37
Specs: ['A1_WDI_only', 'A2_WDI_PCA1', 'A3_WDI_6WGI', 'A4_WDI_PCA_inter', 'A5_WDI_6WGI_inter']
LSTM lookback: 3 anos
Optuna trials: 50


## Célula 4 — Carregar ou extrair os dados (consolidada, CORRIGIDA — imputação não-fold-safe removida)

Esta célula substitui as antigas células 4/4.1/4.2, que duplicavam a mesma lógica
de três formas ligeiramente diferentes. Ordem de tentativa: (1) Drive, (2) disco
local, (3) extracção nova a partir da API do World Bank.

**CORRECTION (achado da conversa: imputação por fold vs. painel completo):** a
versão anterior desta célula ajustava um `PanelMICEImputer` **uma única vez no
painel inteiro** (todos os 37 países × todos os 28 anos, 1996–2023) antes de
qualquer separação treino/teste existir. Isso é um vazamento de look-ahead:
qualquer valor ausente nos indicadores brutos — incluindo, potencialmente, a
própria variável-alvo — era preenchido usando relações estatísticas aprendidas
com o painel inteiro, incluindo anos que seriam "futuro" para os folds mais
antigos. A imputação fold-safe que já existe dentro de `validation/walk_forward.py`
(ajustada de novo, corretamente, só com os anos de treino de cada fold) ficava
então recebendo dados que já tinham sido "vazados" antes — sem muito o que
corrigir.

A correção: esta célula deixa de fazer qualquer imputação MICE no painel
completo. Os `NaN` originais dos dados brutos do World Bank são preservados
em `agregado_inner_join.csv`, e só são preenchidos mais adiante, fold a fold,
pelo `PanelMICEImputer` já existente dentro da Célula 6 (Walk-Forward CV) —
que é o único lugar onde essa imputação deveria acontecer.

**⚠️ Importante:** se você já gerou e guardou `agregado_inner_join.csv` no
Google Drive numa execução anterior, ele foi construído com a imputação antiga
(não-fold-safe) — reexecutar esta célula sozinha vai apenas recarregá-lo do
Drive sem corrigir nada, porque o código de extração nem chega a rodar. Marque
`FORCE_REEXTRACT = True` abaixo **uma vez**, rode a célula para regenerar os
dados brutos sem a imputação antiga, e depois pode voltar a `False`.

In [6]:
import os, sys, time, shutil, warnings
import numpy as np
import pandas as pd
warnings.filterwarnings("ignore")

LOCAL_DIR = "/content/africa_mo_v4"
if LOCAL_DIR not in sys.path:
    sys.path.insert(0, LOCAL_DIR)
os.chdir(LOCAL_DIR)

import config.paths     as paths
import config.variables as var

# CORRECTION: marque True uma vez para forçar a reextracção e descartar
# qualquer agregado_inner_join.csv antigo (gerado com a imputação MICE no
# painel completo, não fold-safe). Depois de regenerar, pode voltar a False.
FORCE_REEXTRACT = False

DRIVE_DATA = "/content/drive/MyDrive/africa_mo_pipeline/data"
os.makedirs(DRIVE_DATA,           exist_ok=True)
os.makedirs(paths.RAW_DIR,        exist_ok=True)
os.makedirs(paths.CLEAN_DIR,      exist_ok=True)
os.makedirs(paths.AGGREGATED_DIR, exist_ok=True)

agg_drive = os.path.join(DRIVE_DATA, "agregado_inner_join.csv")
agg_local = os.path.join(paths.AGGREGATED_DIR, "agregado_inner_join.csv")

df_raw = None

if not FORCE_REEXTRACT and os.path.exists(agg_drive):
    shutil.copy(agg_drive, agg_local)
    df_raw = pd.read_csv(agg_local)
    print(f"✓ Dataset carregado do Drive: {df_raw.shape}")

elif not FORCE_REEXTRACT and os.path.exists(agg_local):
    df_raw = pd.read_csv(agg_local)
    print(f"✓ Dataset carregado do disco local: {df_raw.shape}")

else:
    if FORCE_REEXTRACT:
        print("FORCE_REEXTRACT=True — a ignorar qualquer cache e a extrair dados novos do World Bank...")
    else:
        print("Dataset agregado não encontrado — a extrair dados novos do World Bank...")
    import wbgapi as wb
    anos = range(var.YEAR_START, var.YEAR_END + 1)

    # ── WDI ──────────────────────────────────────────────────────────────────
    print("\nA extrair WDI...")
    dfs = []
    for code, name in var.WDI_INDICATORS.items():
        try:
            df = wb.data.DataFrame(code, var.COUNTRY_CODES, time=anos, labels=False)
            long = df.reset_index().melt(id_vars=["economy"], var_name="year", value_name=name)
            long.rename(columns={"economy": "country_code"}, inplace=True)
            long["year"] = long["year"].astype(str).str.replace("YR", "").astype(int)
            dfs.append(long)
            print(f"  ✓ {name}")
        except Exception as e:
            print(f"  ✗ {code}: {e}")

    df_wdi = dfs[0]
    for d in dfs[1:]:
        df_wdi = df_wdi.merge(d, on=["country_code", "year"], how="outer")
    df_wdi["pais"] = df_wdi["country_code"].map(var.COUNTRIES)
    df_wdi.to_csv(os.path.join(paths.RAW_DIR, "wdi_bruto.csv"), index=False)
    print(f"✓ WDI bruto: {df_wdi.shape}")

    # ── WGI (códigos GOV_WGI_*, via wbgapi db=3) ──────────────────────────────
    print("\nA extrair WGI...")
    WGI_CODES_NOVOS = {
        "GOV_WGI_CC.EST": "wgi_controle_corrupcao",
        "GOV_WGI_GE.EST": "wgi_eficacia_governo",
        "GOV_WGI_PV.EST": "wgi_estabilidade_politica",
        "GOV_WGI_RQ.EST": "wgi_qualidade_regulatoria",
        "GOV_WGI_RL.EST": "wgi_estado_direito",
        "GOV_WGI_VA.EST": "wgi_voz_responsabilizacao",
    }
    wgi_dfs = []
    for code, name in WGI_CODES_NOVOS.items():
        print(f"  {name}...", end=" ", flush=True)
        try:
            wb.db = 3
            df = wb.data.DataFrame(code, economy="all", labels=False)
            wb.db = 2
            long = df.reset_index().melt(id_vars=["economy"], var_name="year", value_name=name)
            long.rename(columns={"economy": "country_code"}, inplace=True)
            long["year"] = pd.to_numeric(
                long["year"].astype(str).str.replace("YR", "").str.strip(), errors="coerce"
            )
            long = long.dropna(subset=["year"])
            long["year"] = long["year"].astype(int)
            long = long[
                long["country_code"].isin(var.COUNTRY_CODES) &
                long["year"].between(var.YEAR_START, var.YEAR_END) &
                long[name].notna()
            ]
            wgi_dfs.append(long[["country_code", "year", name]])
            print(f"✓ {len(long)} registos")
        except Exception as e:
            wb.db = 2
            print(f"✗ {e}")
        time.sleep(1)

    df_wgi = wgi_dfs[0]
    for d in wgi_dfs[1:]:
        df_wgi = df_wgi.merge(d, on=["country_code", "year"], how="outer")
    df_wgi["pais"] = df_wgi["country_code"].map(var.COUNTRIES)
    df_wgi.to_csv(os.path.join(paths.RAW_DIR, "wgi_bruto.csv"), index=False)
    print(f"✓ WGI bruto: {df_wgi.shape}")

    # ── Diagnóstico de valores ausentes (antes de qualquer imputação) ─────────
    print("\nValores ausentes nos dados brutos (serão preenchidos fold-a-fold, não aqui):")
    na_wdi = df_wdi.isna().sum()
    na_wgi = df_wgi.isna().sum()
    print("  WDI:", {k: int(v) for k, v in na_wdi[na_wdi > 0].items()} or "nenhum")
    print("  WGI:", {k: int(v) for k, v in na_wgi[na_wgi > 0].items()} or "nenhum")

    # CORRECTION: a imputação MICE no painel completo foi removida daqui —
    # ajustar um imputador nos 28 anos inteiros de cada país, antes de
    # qualquer split treino/teste, é um vazamento de look-ahead (o imputador
    # fold-safe correcto já existe dentro de validation/walk_forward.py,
    # ajustado apenas com os anos de treino de cada fold). Os ficheiros
    # continuam a chamar-se "*_limpo.csv" só por compatibilidade com o resto
    # do notebook/Drive — mas agora não fazem mais nenhuma limpeza, apenas
    # preservam os NaN originais para serem tratados fold-a-fold mais tarde.
    df_wdi.to_csv(os.path.join(paths.CLEAN_DIR, "wdi_limpo.csv"), index=False)
    df_wgi.to_csv(os.path.join(paths.CLEAN_DIR, "wgi_limpo.csv"), index=False)

    # ── INNER JOIN ─────────────────────────────────────────────────────────────
    df_wdi_c = pd.read_csv(os.path.join(paths.CLEAN_DIR, "wdi_limpo.csv"))
    df_wgi_c = pd.read_csv(os.path.join(paths.CLEAN_DIR, "wgi_limpo.csv")).drop(
        columns=["pais"], errors="ignore"
    )
    df_raw = pd.merge(df_wdi_c, df_wgi_c, on=["country_code", "year"],
                       how="inner", validate="1:1")
    df_raw.to_csv(agg_local, index=False)

    # ── Backup no Drive ────────────────────────────────────────────────────────
    for fname in ["wdi_bruto.csv", "wgi_bruto.csv"]:
        shutil.copy(os.path.join(paths.RAW_DIR, fname), DRIVE_DATA)
    for fname in ["wdi_limpo.csv", "wgi_limpo.csv"]:
        shutil.copy(os.path.join(paths.CLEAN_DIR, fname), DRIVE_DATA)
    shutil.copy(agg_local, agg_drive)
    print(f"✓ Dataset extraído e guardado no Drive: {df_raw.shape}")

print(f"\n  Países : {df_raw['country_code'].nunique()}")
print(f"  Anos   : {df_raw['year'].min()}–{df_raw['year'].max()}")
print(f"  Target NaN: {df_raw['valor_agregado_industrial_percent_pib'].isna().sum()} / {len(df_raw)}")
na_cols = df_raw.isna().sum()
na_cols = na_cols[na_cols > 0]
if len(na_cols):
    print(f"  Colunas com NaN pendente (serão preenchidas fold-a-fold na Célula 6): {dict(na_cols)}")
else:
    print("  Nenhum NaN pendente no painel agregado.")


✓ Dataset carregado do Drive: (925, 28)

  Países : 37
  Anos   : 1996–2023
  Target NaN: 39 / 925
  Colunas com NaN pendente (serão preenchidas fold-a-fold na Célula 6): {'valor_agregado_industrial_percent_pib': np.int64(39), 'pib_per_capita_ppc': np.int64(25), 'crescimento_pib_anual': np.int64(5), 'formacao_bruta_capital_fixo_percent_pib': np.int64(150), 'ied_percent_pib': np.int64(13), 'comercio_percent_pib': np.int64(65), 'exportacoes_percent_pib': np.int64(65), 'importacoes_percent_pib': np.int64(65), 'utilizadores_internet_percent': np.int64(17), 'despesa_id_percent_pib': np.int64(643), 'despesa_educacao_percent_pib': np.int64(349), 'taxa_matricula_terciario': np.int64(317), 'credito_privado_percent_pib': np.int64(166), 'valor_agregado_manufatura_percent_pib': np.int64(59), 'rendas_recursos_naturais_percent_pib': np.int64(78)}


**Opcional — recuperar modelos já treinados do Drive** (útil se a sessão do
Colab foi reiniciada e não quer re-treinar do zero).

In [ ]:
import os, shutil
import config.paths as paths

DRIVE_MODELS = "/content/drive/MyDrive/africa_mo_pipeline/models/artefacts"
os.makedirs(paths.MODELS_DIR, exist_ok=True)

if os.path.exists(DRIVE_MODELS):
    pkls = [f for f in os.listdir(DRIVE_MODELS) if f.endswith(".pkl")]
    for f in pkls:
        shutil.copy(os.path.join(DRIVE_MODELS, f), os.path.join(paths.MODELS_DIR, f))
    print(f"✓ {len(pkls)} modelos carregados do Drive")
else:
    print("Nenhum modelo no Drive ainda — normal na primeira execução.")


## Célula 6 — Treinamento Walk-Forward CV (2–3 horas com GPU)

In [7]:
import os, sys
LOCAL_DIR = "/content/africa_mo_v4"
if LOCAL_DIR not in sys.path:
    sys.path.insert(0, LOCAL_DIR)
os.chdir(LOCAL_DIR)

import pickle
import numpy as np
import pandas as pd
import config.paths        as paths
import config.features     as feat
import config.model_params as mp
import config.pipeline     as cfg_pipe

from validation.walk_forward import WalkForwardCV
from models.rf.model         import train as train_rf
from models.xgb.model        import train as train_xgb
from models.sarimax.model    import train as train_sarimax
from models.lstm.model       import train as train_lstm
from models.bayesian.model   import train as train_bayesian

if "df_raw" not in dir() or df_raw is None:
    df_raw = pd.read_csv(os.path.join(paths.AGGREGATED_DIR, "agregado_inner_join.csv"))
    print(f"df_raw carregado: {df_raw.shape}")

MODEL_TRAINERS = {
    "RandomForest":   train_rf,
    "XGBoost":        train_xgb,
    "SARIMAX":        train_sarimax,
    "LSTM":           train_lstm,
    "Bayes_Partial":  lambda Xt, yt, Xv, yv: train_bayesian(Xt, yt, Xv, yv, "partial"),
    "Bayes_Complete": lambda Xt, yt, Xv, yv: train_bayesian(Xt, yt, Xv, yv, "complete"),
}

wf          = WalkForwardCV()
all_results = []
hp_records  = []
os.makedirs(paths.MODELS_DIR, exist_ok=True)

for spec in feat.ABLATION_SPECS:
    print(f"\n{'─'*55}")
    print(f"Especificação: {spec}")
    for mod_name, trainer_fn in MODEL_TRAINERS.items():
        print(f"  [{mod_name}]")
        try:
            # NOTA: wf.evaluate() já grava, no fim, um bundle {model, scaler,
            # feat_cols} em vez do modelo isolado (correcção #3) — nada a
            # mudar aqui, a correcção está dentro de walk_forward.py.
            fold_res = wf.evaluate(df_raw, spec, trainer_fn, mod_name, save_model=True)
            rmse_vals = [r.RMSE for r in fold_res if not np.isnan(r.RMSE)]
            mean_rmse = np.mean(rmse_vals) if rmse_vals else np.nan
            all_results.extend([
                {"spec": r.spec, "model": r.model, "fold": r.fold,
                 "RMSE": r.RMSE, "MAE": r.MAE, "R2": r.R2, "MASE": r.MASE}
                for r in fold_res
            ])
            hp_records.append({
                "Specification": spec, "Model": mod_name,
                "Search": "Optuna TPE", "Seed": mp.SEED,
                "Mean_RMSE_WF": round(mean_rmse, 4) if not np.isnan(mean_rmse) else None,
            })
            print(f"    RMSE médio = {mean_rmse:.4f}" if not np.isnan(mean_rmse) else "    todos os folds falharam")
        except Exception as e:
            import traceback
            print(f"    ERRO: {e}")
            traceback.print_exc()

df_results = pd.DataFrame(all_results)
os.makedirs(paths.REPORTS_DIR, exist_ok=True)
results_path = os.path.join(paths.REPORTS_DIR, "walkforward_results.csv")
df_results.to_csv(results_path, index=False)

pkls = [f for f in os.listdir(paths.MODELS_DIR) if f.endswith(".pkl")]
print(f"\n✓ Modelos guardados (bundle model+scaler+feat_cols): {len(pkls)}")
for f in sorted(pkls):
    kb = os.path.getsize(os.path.join(paths.MODELS_DIR, f)) / 1024
    print(f"  {f}  ({kb:.0f} KB)")

print(f"\n✓ Resultados: {results_path}")
if not df_results.dropna(subset=["RMSE"]).empty:
    print(df_results.groupby(["spec", "model"])["RMSE"].mean().unstack().round(4).to_string())



───────────────────────────────────────────────────────
Especificação: A1_WDI_only
  [RandomForest]
      Fold 1/5 — RMSE=3.7776  R²=0.9555
      Fold 2/5 — RMSE=2.9026  R²=0.9672
      Fold 3/5 — RMSE=4.5913  R²=0.8171
      Fold 4/5 — RMSE=2.5041  R²=0.9545
      Fold 5/5 — RMSE=4.6079  R²=0.8346
    RMSE médio = 3.6767
  [XGBoost]
      Fold 1/5 — RMSE=4.3067  R²=0.9421
      Fold 2/5 — RMSE=2.5792  R²=0.9741
      Fold 3/5 — RMSE=3.6053  R²=0.8872
      Fold 4/5 — RMSE=2.9051  R²=0.9387
      Fold 5/5 — RMSE=4.4689  R²=0.8444
    RMSE médio = 3.5731
  [SARIMAX]
      Fold 1/5 — RMSE=2.6508  R²=0.9781
      Fold 2/5 — RMSE=2.1423  R²=0.9821
      Fold 3/5 — RMSE=2.3115  R²=0.9536
      Fold 4/5 — RMSE=2.0527  R²=0.9694
      Fold 5/5 — RMSE=3.2444  R²=0.9180
    RMSE médio = 2.4803
  [LSTM]
      Fold 1/5 — RMSE=7.7782  R²=0.8120
      Fold 2/5 — RMSE=4.3548  R²=0.9269
      Fold 3/5 — RMSE=6.0022  R²=0.6930
      Fold 4/5 — RMSE=3.9882  R²=0.8865
      Fold 5/5 — RMSE=3.8407  R²=0

**Verificação da existência dos ficheiros (opcional, útil para depurar)**

In [8]:
import os, sys
LOCAL_DIR = "/content/africa_mo_v4"
if LOCAL_DIR not in sys.path:
    sys.path.insert(0, LOCAL_DIR)
os.chdir(LOCAL_DIR)

import pandas as pd
import numpy as np
import config.paths     as paths
import config.features  as feat

print("=" * 55)
print("1. DATASET")
print("=" * 55)
if "df_raw" not in dir():
    agg_path = os.path.join(paths.AGGREGATED_DIR, "agregado_inner_join.csv")
    if os.path.exists(agg_path):
        df_raw = pd.read_csv(agg_path)
        print(f"  df_raw carregado do disco: {df_raw.shape}")
    else:
        raise RuntimeError("Executar a Célula 4 primeiro")
else:
    print(f"  df_raw em memória: {df_raw.shape}")

print("\n" + "=" * 55)
print("2. TESTE WALK-FORWARD (1 spec x 1 modelo x 1 fold)")
print("=" * 55)
try:
    from validation.walk_forward import WalkForwardCV
    from models.rf.model import train as train_rf

    wf   = WalkForwardCV(n_folds=1)
    # CORRECTION: usar a chave real de ABLATION_SPECS (era "WDI_only", que
    # não existe — devia ser "A1_WDI_only").
    spec = "A1_WDI_only"

    print(f"  Spec: {spec}")
    fold_res = wf.evaluate(df_raw, spec, train_rf, "RandomForest")
    for r in fold_res:
        print(f"  Fold {r.fold}: RMSE={r.RMSE:.4f}  R²={r.R2:.4f}  n_train={r.n_train}  n_test={r.n_test}")
    print("  ✓ Walk-forward funciona")
except Exception:
    import traceback
    traceback.print_exc()

print("\n" + "=" * 55)
print("3. MODELOS GUARDADOS")
print("=" * 55)
if os.path.exists(paths.MODELS_DIR):
    pkls = sorted(f for f in os.listdir(paths.MODELS_DIR) if f.endswith(".pkl"))
    for f in pkls:
        kb = os.path.getsize(os.path.join(paths.MODELS_DIR, f)) / 1024
        print(f"  ✓ {f}  ({kb:.0f} KB)")
    if not pkls:
        print("  Nenhum modelo .pkl guardado ainda")
else:
    print(f"  Pasta não existe: {paths.MODELS_DIR}")


1. DATASET
  df_raw em memória: (925, 28)

2. TESTE WALK-FORWARD (1 spec x 1 modelo x 1 fold)
  Spec: A1_WDI_only
      Fold 1/1 — RMSE=4.6097  R²=0.8978
  Fold 1: RMSE=4.6097  R²=0.8978  n_train=346  n_test=481
  ✓ Walk-forward funciona

3. MODELOS GUARDADOS
  ✓ modelo_A1_WDI_only_Bayes_Complete.pkl  (19646 KB)
  ✓ modelo_A1_WDI_only_Bayes_Partial.pkl  (19708 KB)
  ✓ modelo_A1_WDI_only_LSTM.pkl  (169 KB)
  ✓ modelo_A1_WDI_only_RandomForest.pkl  (540 KB)
  ✓ modelo_A1_WDI_only_SARIMAX.pkl  (46 KB)
  ✓ modelo_A1_WDI_only_XGBoost.pkl  (143 KB)
  ✓ modelo_A2_WDI_PCA1_Bayes_Complete.pkl  (19646 KB)
  ✓ modelo_A2_WDI_PCA1_Bayes_Partial.pkl  (19709 KB)
  ✓ modelo_A2_WDI_PCA1_LSTM.pkl  (173 KB)
  ✓ modelo_A2_WDI_PCA1_RandomForest.pkl  (2218 KB)
  ✓ modelo_A2_WDI_PCA1_SARIMAX.pkl  (46 KB)
  ✓ modelo_A2_WDI_PCA1_XGBoost.pkl  (350 KB)
  ✓ modelo_A3_WDI_6WGI_Bayes_Complete.pkl  (19646 KB)
  ✓ modelo_A3_WDI_6WGI_Bayes_Partial.pkl  (19709 KB)
  ✓ modelo_A3_WDI_6WGI_LSTM.pkl  (182 KB)
  ✓ modelo_A3_

## Célula 6.1 — Treinar apenas os modelos Bayesianos com MCMC (opcional)

Esta secção substitui o fallback rápido (BayesianRidge) pelo MCMC real do PyMC.
Corra-a **numa sessão separada**, porque exige uma combinação de numpy/PyMC
diferente da usada nas células anteriores — depois de instalar, é preciso
reiniciar o runtime (`Runtime → Restart session`) antes de treinar.

In [9]:
import subprocess
print("A instalar numpy>=2.0, PyMC 5.x e dependências compatíveis...")
subprocess.run([
    "pip", "install", "-q",
    "numpy>=2.0", "pymc>=5.0", "pytensor", "arviz>=0.17", "scikit-learn>=1.5",
], check=True)
print("✓ Instalação concluída.")
print("\nATENÇÃO: reinicie o runtime agora (Runtime → Restart session).")
print("Depois de reiniciar, corra apenas a célula seguinte (treino MCMC) —")
print("não volte a correr as células de Walk-Forward CV nesta mesma sessão.")


A instalar numpy>=2.0, PyMC 5.x e dependências compatíveis...
✓ Instalação concluída.

ATENÇÃO: reinicie o runtime agora (Runtime → Restart session).
Depois de reiniciar, corra apenas a célula seguinte (treino MCMC) —
não volte a correr as células de Walk-Forward CV nesta mesma sessão.


In [10]:
import os, sys
LOCAL_DIR = "/content/africa_mo_v4"
if LOCAL_DIR not in sys.path:
    sys.path.insert(0, LOCAL_DIR)
os.chdir(LOCAL_DIR)

import numpy, sklearn, pymc
print(f"numpy:   {numpy.__version__}")
print(f"sklearn: {sklearn.__version__}")
print(f"pymc:    {pymc.__version__}")

import pickle
import numpy as np
import pandas as pd
import config.paths        as paths
import config.features     as feat
from validation.walk_forward import WalkForwardCV
from models.bayesian.model   import train as train_bayesian

df_raw = pd.read_csv(os.path.join(paths.AGGREGATED_DIR, "agregado_inner_join.csv"))
print(f"\n✓ Dataset: {df_raw.shape}")

MODEL_TRAINERS = {
    "Bayes_Partial_MCMC":  lambda Xt, yt, Xv, yv: train_bayesian(Xt, yt, Xv, yv, "partial"),
    "Bayes_Complete_MCMC": lambda Xt, yt, Xv, yv: train_bayesian(Xt, yt, Xv, yv, "complete"),
}

wf, all_results = WalkForwardCV(), []
os.makedirs(paths.MODELS_DIR, exist_ok=True)

for spec in feat.ABLATION_SPECS:
    print(f"\n── {spec} ──")
    for mod_name, trainer_fn in MODEL_TRAINERS.items():
        print(f"  [{mod_name}]")
        try:
            fold_res = wf.evaluate(df_raw, spec, trainer_fn, mod_name, save_model=True)
            rmse_vals = [r.RMSE for r in fold_res if not np.isnan(r.RMSE)]
            print(f"    RMSE médio = {np.mean(rmse_vals):.4f}" if rmse_vals else "    sem resultados")
            all_results.extend([
                {"spec": r.spec, "model": r.model, "fold": r.fold,
                 "RMSE": r.RMSE, "MAE": r.MAE, "R2": r.R2, "MASE": r.MASE}
                for r in fold_res
            ])
        except Exception as e:
            print(f"    ERRO: {e}")

df_bayes = pd.DataFrame(all_results)
bayes_path = os.path.join(paths.REPORTS_DIR, "walkforward_bayesian_mcmc.csv")
df_bayes.to_csv(bayes_path, index=False)

prev_path = os.path.join(paths.REPORTS_DIR, "walkforward_results.csv")
if os.path.exists(prev_path):
    df_todos = pd.concat([pd.read_csv(prev_path), df_bayes], ignore_index=True)
    df_todos.to_csv(os.path.join(paths.REPORTS_DIR, "walkforward_todos_modelos.csv"), index=False)
    print("\n── Comparação completa ──")
    print(df_todos.groupby(["spec", "model"])["RMSE"].mean().unstack().round(4).to_string())


numpy:   2.0.2
sklearn: 1.6.1
pymc:    5.28.5

✓ Dataset: (925, 28)

── A1_WDI_only ──
  [Bayes_Partial_MCMC]
      Fold 1/5 — RMSE=14.3945  R²=0.3533
      Fold 2/5 — RMSE=11.9703  R²=0.4414
      Fold 3/5 — RMSE=10.6097  R²=0.0233
      Fold 4/5 — RMSE=9.8237  R²=0.2994
      Fold 5/5 — RMSE=10.5834  R²=0.1276
    RMSE médio = 11.4763
  [Bayes_Complete_MCMC]
      Fold 1/5 — RMSE=14.6506  R²=0.3301
      Fold 2/5 — RMSE=12.2119  R²=0.4186
      Fold 3/5 — RMSE=10.6592  R²=0.0141
      Fold 4/5 — RMSE=9.7690  R²=0.3071
      Fold 5/5 — RMSE=10.1757  R²=0.1935
    RMSE médio = 11.4933

── A2_WDI_PCA1 ──
  [Bayes_Partial_MCMC]
    PCA fitted on 444 obs — PC1 explains 79.8% variance
      Fold 1/5 — RMSE=14.3945  R²=0.3533
    PCA fitted on 518 obs — PC1 explains 79.8% variance
      Fold 2/5 — RMSE=11.9703  R²=0.4414
    PCA fitted on 592 obs — PC1 explains 79.6% variance
      Fold 3/5 — RMSE=10.6097  R²=0.0233
    PCA fitted on 666 obs — PC1 explains 79.6% variance
      Fold 4/5 — RM

**Ver todos os ficheiros de modelo treinados**

In [11]:
import os
import config.paths as paths
pkls = sorted(f for f in os.listdir(paths.MODELS_DIR) if f.endswith(".pkl"))
for f in pkls:
    print(f)


modelo_A1_WDI_only_Bayes_Complete.pkl
modelo_A1_WDI_only_Bayes_Complete_MCMC.pkl
modelo_A1_WDI_only_Bayes_Partial.pkl
modelo_A1_WDI_only_Bayes_Partial_MCMC.pkl
modelo_A1_WDI_only_LSTM.pkl
modelo_A1_WDI_only_RandomForest.pkl
modelo_A1_WDI_only_SARIMAX.pkl
modelo_A1_WDI_only_XGBoost.pkl
modelo_A2_WDI_PCA1_Bayes_Complete.pkl
modelo_A2_WDI_PCA1_Bayes_Complete_MCMC.pkl
modelo_A2_WDI_PCA1_Bayes_Partial.pkl
modelo_A2_WDI_PCA1_Bayes_Partial_MCMC.pkl
modelo_A2_WDI_PCA1_LSTM.pkl
modelo_A2_WDI_PCA1_RandomForest.pkl
modelo_A2_WDI_PCA1_SARIMAX.pkl
modelo_A2_WDI_PCA1_XGBoost.pkl
modelo_A3_WDI_6WGI_Bayes_Complete.pkl
modelo_A3_WDI_6WGI_Bayes_Complete_MCMC.pkl
modelo_A3_WDI_6WGI_Bayes_Partial.pkl
modelo_A3_WDI_6WGI_Bayes_Partial_MCMC.pkl
modelo_A3_WDI_6WGI_LSTM.pkl
modelo_A3_WDI_6WGI_RandomForest.pkl
modelo_A3_WDI_6WGI_SARIMAX.pkl
modelo_A3_WDI_6WGI_XGBoost.pkl
modelo_A4_WDI_PCA_inter_Bayes_Complete.pkl
modelo_A4_WDI_PCA_inter_Bayes_Complete_MCMC.pkl
modelo_A4_WDI_PCA_inter_Bayes_Partial.pkl
modelo_A4

## Célula 7 — SHAP + Permutation importance (CORRIGIDA)

**Esta era a célula onde vivia o problema mais consequente de todo o pipeline**
(achado nº7 do relatório de causas-raiz): `X_all`/`X_test` eram construídos com
os valores brutos das colunas e alimentados directamente a `shap_tree_analysis()`
e `permutation_importance()` — mas os modelos foram treinados sobre dados
estandardizados (`StandardScaler`). Isso distorcia sistematicamente as
importâncias reportadas. Este bug era **específico do notebook** (não estava
apenas no `pipeline.py`), por isso continua reescrito aqui mesmo nesta versão
enxuta, mesmo que o resto das correções venha do repositório GitHub.

A célula abaixo carrega, para cada modelo, o *bundle* `{model, scaler,
feat_cols}` agora gravado pela Célula 6 (graças à correcção em
`validation/walk_forward.py`), alinha as colunas pela mesma ordem usada no
treino, e aplica `scaler.transform()` antes de chamar o SHAP e a permutação.

In [12]:
import os, pickle
import numpy as np
import pandas as pd
from features.engineer            import FoldFeatureEngineer
from explainability.shap_analysis import shap_tree_analysis
from explainability.permutation   import permutation_importance
from utils.model_io               import load_model_bundle
import config.paths     as paths
import config.variables as var
import config.pipeline  as cfg_pipe

years     = sorted(df_raw["year"].unique())
split_idx = int(len(years) * (1 - cfg_pipe.FINAL_HOLDOUT_RATIO))
test_yr   = years[split_idx:]

SPECS = ["A1_WDI_only", "A2_WDI_PCA1", "A3_WDI_6WGI", "A4_WDI_PCA_inter", "A5_WDI_6WGI_inter"]
MODELOS_TREE = ["RandomForest", "XGBoost"]

for spec in SPECS:
    print(f"\n── {spec} ──")
    fe = FoldFeatureEngineer(spec=spec)
    fe.fit(df_raw)
    df_fe = fe.transform(df_raw)
    feat_cols_local = [c for c in df_fe.select_dtypes(include=[np.number]).columns
                        if c not in {"year", var.TARGET} and "country" not in c.lower()]

    for mod in MODELOS_TREE:
        pkl = os.path.join(paths.MODELS_DIR, f"modelo_{spec}_{mod}.pkl")
        if not os.path.exists(pkl):
            print(f"  [{mod}] não encontrado")
            continue

        # CORRECTION: carregar o bundle (modelo + scaler + colunas de treino)
        # em vez do modelo isolado, e usar exactamente as colunas/ordem do treino.
        model, scaler, trained_cols = load_model_bundle(pkl)
        cols = [c for c in (trained_cols or feat_cols_local) if c in df_fe.columns]

        X_all_raw  = df_fe[cols].fillna(0)
        X_test_raw = df_fe[df_fe["year"].isin(test_yr)][cols].fillna(0)
        y_test     = df_fe[df_fe["year"].isin(test_yr)][var.TARGET].values

        if scaler is not None:
            X_all  = pd.DataFrame(scaler.transform(X_all_raw.values),  columns=cols, index=X_all_raw.index)
            X_test = pd.DataFrame(scaler.transform(X_test_raw.values), columns=cols, index=X_test_raw.index)
        else:
            print(f"    [aviso] {spec}_{mod}: pickle sem scaler (formato anterior à correcção) — usando dados brutos.")
            X_all, X_test = X_all_raw, X_test_raw

        label = f"{spec}_{mod}"
        print(f"  [{mod}] SHAP...", end=" ", flush=True)
        try:
            r = shap_tree_analysis(model, X_all, label, paths.EXPLAINABILITY_DIR)
            print(f"OK (governance {r.get('gov_pct', 0):.1f}%)", end="  ")
        except Exception as e:
            print(f"ERRO: {e}", end="  ")
        print("Permutation...", end=" ", flush=True)
        try:
            permutation_importance(model, X_test, y_test, label, paths.EXPLAINABILITY_DIR)
            print("OK")
        except Exception as e:
            print(f"ERRO: {e}")

print(f"\n✓ Ficheiros em: {paths.EXPLAINABILITY_DIR}")
ficheiros = [f for f in sorted(os.listdir(paths.EXPLAINABILITY_DIR))
             if os.path.isfile(os.path.join(paths.EXPLAINABILITY_DIR, f))]
print(f"  Total: {len(ficheiros)} ficheiros")



── A1_WDI_only ──
  [RandomForest] SHAP... OK (governance 0.0%)  Permutation... OK
  [XGBoost] SHAP... OK (governance 0.0%)  Permutation... OK

── A2_WDI_PCA1 ──
    PCA fitted on 925 obs — PC1 explains 80.0% variance
  [RandomForest] SHAP... OK (governance 0.4%)  Permutation... OK
  [XGBoost] SHAP... OK (governance 0.8%)  Permutation... OK

── A3_WDI_6WGI ──
  [RandomForest] SHAP... OK (governance 1.5%)  Permutation... OK
  [XGBoost] SHAP... OK (governance 3.3%)  Permutation... OK

── A4_WDI_PCA_inter ──
    PCA fitted on 925 obs — PC1 explains 80.0% variance
  [RandomForest] SHAP... OK (governance 0.7%)  Permutation... OK
  [XGBoost] SHAP... OK (governance 1.2%)  Permutation... OK

── A5_WDI_6WGI_inter ──
  [RandomForest] SHAP... OK (governance 0.9%)  Permutation... OK
  [XGBoost] SHAP... OK (governance 2.0%)  Permutation... OK

✓ Ficheiros em: /content/africa_mo_v4/explainability/results
  Total: 98 ficheiros


## Célula 8 — Ablação (responde à hipótese WGI)

`run_ablation()` já contém, depois da correcção, a leitura do *bundle* de
modelo (não apenas o modelo isolado) e uma especificação-baseline explícita
(`A1_WDI_only`, em vez de "o primeiro item do dicionário") — nada precisa de
ser alterado nesta célula, que já chamava a função correctamente. Se a Célula
2.1 acusou ✗ neste ficheiro, actualize o GitHub antes de continuar.

In [13]:
import os, shutil
from explainability.ablation import run_ablation
from features.engineer import FoldFeatureEngineer
import config.features as feat
import config.paths as paths

MODELOS_ABLACAO = ["RandomForest", "XGBoost", "SARIMAX", "LSTM", "Bayes_Partial", "Bayes_Complete"]

spec_datasets = {}
for spec in feat.ABLATION_SPECS:
    fe = FoldFeatureEngineer(spec=spec)
    fe.fit(df_raw)
    spec_datasets[spec] = fe.transform(df_raw)
    print(f"  {spec}: {spec_datasets[spec].shape}")

years     = sorted(df_raw["year"].unique())
import config.pipeline as cfg_pipe
split_idx = int(len(years) * (1 - cfg_pipe.FINAL_HOLDOUT_RATIO))

abl_dir = os.path.join(paths.EXPLAINABILITY_DIR, "ablation")
df_abl  = run_ablation(
    spec_datasets=spec_datasets,
    model_names=MODELOS_ABLACAO,
    model_dir=paths.MODELS_DIR,
    out_dir=abl_dir,
    test_year_cutoff=years[split_idx],
)

if not df_abl.empty:
    print("\nRMSE por especificação x modelo:")
    print(df_abl.pivot_table(values="RMSE", index="Specification", columns="Model", aggfunc="mean").round(4).to_string())
    print("\nRMSE médio por especificação:")
    print(df_abl.groupby("Specification")[["RMSE", "R2"]].mean().round(4))

    if paths.DRIVE_DIR:
        shutil.copytree(abl_dir, os.path.join(paths.DRIVE_DIR, "explainability", "ablation"), dirs_exist_ok=True)
    os.system(f"cd {LOCAL_DIR} && git add explainability/ && git commit -m 'ablation results - all specs' && git push")
    print(f"\n✓ Guardado em: {abl_dir}")
else:
    print("✗ Nenhum resultado")


  A1_WDI_only: (888, 47)
    PCA fitted on 925 obs — PC1 explains 80.0% variance
  A2_WDI_PCA1: (888, 51)
  A3_WDI_6WGI: (888, 59)
    PCA fitted on 925 obs — PC1 explains 80.0% variance
  A4_WDI_PCA_inter: (888, 55)
  A5_WDI_6WGI_inter: (888, 63)
  Saved: /content/africa_mo_v4/explainability/results/ablation/ablation_results.csv
  Saved: /content/africa_mo_v4/explainability/results/ablation/ablation_dm_tests.csv
  Saved: /content/africa_mo_v4/explainability/results/ablation/ablation_rmse_heatmap.png
  Saved: /content/africa_mo_v4/explainability/results/ablation/ablation_rmse_gain.png

  Ablation: 30 model×spec combinations
  DM tests significant at 5%: 24/24

RMSE por especificação x modelo:
Model              Bayes_Complete  Bayes_Partial    LSTM  RandomForest  SARIMAX  XGBoost
Specification                                                                           
A1_WDI_only               12.2495        12.4653  6.8476        6.6933   7.7623   6.7982
A2_WDI_PCA1               12.24

**Guardar modelos no Drive**

In [14]:
import shutil
DRIVE_MODELS = "/content/drive/MyDrive/africa_mo_pipeline/models/artefacts"
os.makedirs(DRIVE_MODELS, exist_ok=True)
for f in os.listdir(paths.MODELS_DIR):
    if f.endswith(".pkl"):
        shutil.copy(os.path.join(paths.MODELS_DIR, f), DRIVE_MODELS)
print(f"✓ {len(os.listdir(DRIVE_MODELS))} modelos guardados no Drive")


✓ 40 modelos guardados no Drive


# Quebras estruturais e regimes, Contrafactual e ACI (Adaptive Conformal Inference)

A célula que antes procurava `innovations.py` em vários locais do Drive/Colab
foi removida — o ficheiro correcto (`explainability/innovations.py`) vem do
repositório GitHub actualizado (verificado na Célula 2.1). `run_all_innovations()`
já aplica o scaler persistido antes de chamar `predict()` no contrafactual e no
ACI (correcção #3).

In [15]:
from explainability.innovations import run_all_innovations
import os

results = run_all_innovations(df_raw, spec="A2_WDI_PCA1")

os.system(f"""
cd {LOCAL_DIR} && \
git add explainability/ && \
git commit -m "innovations: structural breaks + counterfactual + ACI (corrigido)" && \
git push
""")
print("✓ GitHub actualizado")


  A construir features para spec=A2_WDI_PCA1...
    PCA fitted on 925 obs — PC1 explains 80.0% variance
  ✓ Dataset com features: (888, 51)

  INNOVATION 1: STRUCTURAL BREAKS + REGIMES
  ruptures disponível — usando PELT
  Breaks detected: 78
  Regimes classified: 886 country×year observations

  INNOVATION 2: COUNTERFACTUAL SIMULATION
  Counterfactual simulations: 740

  INNOVATION 3: ACI SENSITIVITY ANALYSIS
    γ=0.01   w=3     coverage=69.3%  width=5.640
    γ=0.01   w=5     coverage=74.9%  width=6.590
    γ=0.01   w=7     coverage=77.1%  width=7.119
    γ=0.01   w=10    coverage=78.2%  width=7.362
    γ=0.01   w=15    coverage=79.3%  width=7.720
    γ=0.05   w=3     coverage=68.7%  width=5.596
    γ=0.05   w=5     coverage=74.3%  width=6.524
    γ=0.05   w=7     coverage=77.1%  width=7.022
    γ=0.05   w=10    coverage=78.2%  width=7.227
    γ=0.05   w=15    coverage=79.3%  width=7.638
    γ=0.1    w=3     coverage=67.6%  width=5.535
    γ=0.1    w=5     coverage=73.7%  width=6.43

## Célula 9 — Relatórios + commit GitHub (CORRIGIDA)

O relatório de causas-raiz apontou que `best_RMSE` (o mínimo entre as 150
linhas de fold individuais) e a tabela de desempenho agregada (média por
especificação×modelo) são duas estatísticas diferentes do mesmo
`walkforward_results.csv`, mas o resumo executivo só mostrava uma delas sem
rótulo — parecendo uma inconsistência. Esta célula calcula e rotula as duas
explicitamente.

In [16]:
import os, shutil
import pandas as pd
from reports.report_generator import run_all_reports

results_path = os.path.join(paths.REPORTS_DIR, "walkforward_results.csv")
abl_dir      = os.path.join(paths.EXPLAINABILITY_DIR, "ablation")

df_results = pd.read_csv(results_path) if os.path.exists(results_path) else pd.DataFrame()

# CORRECTION: reportar as duas estatísticas, cada uma com rótulo explícito,
# em vez de apenas "best_RMSE" (que era o mínimo de um único fold, sem dizer isso).
summary_kv = {"n_models": 6, "n_specs": 5, "n_folds": 5}
if not df_results.empty and "RMSE" in df_results.columns:
    best_single = df_results.loc[df_results["RMSE"].idxmin()]
    df_mean_grp = df_results.groupby(["spec", "model"])["RMSE"].mean().reset_index()
    best_mean   = df_mean_grp.loc[df_mean_grp["RMSE"].idxmin()]

    summary_kv.update({
        "best_RMSE_single_fold (mínimo entre as linhas de fold individuais)": round(float(best_single["RMSE"]), 4),
        "best_model_single_fold": str(best_single["model"]),
        "best_spec_single_fold":  str(best_single["spec"]),
        "best_RMSE_mean_per_group (média por especificação×modelo — comparável à tabela de desempenho)": round(float(best_mean["RMSE"]), 4),
        "best_model_mean_per_group": str(best_mean["model"]),
        "best_spec_mean_per_group":  str(best_mean["spec"]),
    })

abl_csv    = os.path.join(abl_dir, "ablation_results.csv")
abl_dm_csv = os.path.join(abl_dir, "ablation_dm_tests.csv")

run_all_reports(
    results_csv     = results_path if os.path.exists(results_path) else None,
    hp_csv          = None,
    ablation_csv    = abl_csv    if os.path.exists(abl_csv)    else None,
    ablation_dm_csv = abl_dm_csv if os.path.exists(abl_dm_csv) else None,
    summary_kv      = summary_kv,
)

DRIVE_OUT = "/content/drive/MyDrive/africa_mo_pipeline/"
for folder in ["reports", "explainability"]:
    src = os.path.join(paths.ROOT, folder)
    if os.path.exists(src):
        shutil.copytree(src, os.path.join(DRIVE_OUT, folder), dirs_exist_ok=True)
        print(f"✓ Drive ← {folder}/")

os.system(f"""
cd {LOCAL_DIR} && \
git add reports/ explainability/ && \
git commit -m "reports: walk-forward + ablation (corrigido) $(date +%Y-%m-%d)" && \
git push
""")
print("✓ Concluído.")
print("\nFicheiros em reports/:")
for f in sorted(os.listdir(paths.REPORTS_DIR)):
    print(f"  {f}")



  REPORTS: generating dissertation tables
  [report] Performance table → /content/africa_mo_v4/reports/table_performance.csv  (30 rows)
  [report] Ablation table → /content/africa_mo_v4/reports/table_ablation.csv  (30 rows)
  [report] Executive summary → /content/africa_mo_v4/reports/executive_summary.md

  5 report files generated in /content/africa_mo_v4/reports/
✓ Drive ← reports/
✓ Drive ← explainability/
✓ Concluído.

Ficheiros em reports/:
  __init__.py
  __pycache__
  executive_summary.md
  report_generator.py
  table_ablation.csv
  table_ablation.tex
  table_performance.csv
  table_performance.tex
  walkforward_bayesian_mcmc.csv
  walkforward_results.csv
  walkforward_todos_modelos.csv


**Guardar relatório no Git e no Drive**

In [17]:
import shutil, os
DRIVE_OUT = "/content/drive/MyDrive/africa_mo_pipeline/"
shutil.copytree(paths.REPORTS_DIR, os.path.join(DRIVE_OUT, "reports"), dirs_exist_ok=True)
print("✓ Drive ← reports/")

os.system(f"""
cd {LOCAL_DIR} && \
git add reports/ && \
git commit -m "reports: performance + ablation tables $(date +%Y-%m-%d)" && \
git push
""")
print("✓ GitHub actualizado")


✓ Drive ← reports/
✓ GitHub actualizado


**Listar todos os ficheiros gerados**

In [18]:
import os

LOCAL_DIR = "/content/africa_mo_v4"
DRIVE_DIR = "/content/drive/MyDrive/africa_mo_pipeline"

def listar_ficheiros(raiz, label):
    print(f"\n{'='*60}")
    print(f"  {label}")
    print(f"  {raiz}")
    print(f"{'='*60}")
    if not os.path.exists(raiz):
        print("  ✗ Pasta não existe")
        return 0
    total, tamanho_total = 0, 0
    for dirpath, dirnames, filenames in os.walk(raiz):
        dirnames[:] = [d for d in dirnames if d != "__pycache__"]
        ficheiros = [f for f in filenames if not f.endswith(".pyc")]
        if not ficheiros:
            continue
        pasta_rel = os.path.relpath(dirpath, raiz)
        if pasta_rel == ".":
            pasta_rel = ""
        print(f"\n  📁 {pasta_rel or '(raiz)'}/")
        for f in sorted(ficheiros):
            caminho = os.path.join(dirpath, f)
            kb = os.path.getsize(caminho) / 1024
            print(f"     {f}  ({kb:.0f} KB)")
            total += 1
            tamanho_total += kb
    print(f"\n  Total: {total} ficheiros  |  {tamanho_total/1024:.1f} MB")
    return total

PASTAS_COLAB = [
    (os.path.join(LOCAL_DIR, "data"),                    "DADOS (raw, clean, aggregated, features)"),
    (os.path.join(LOCAL_DIR, "models/artefacts"),        "MODELOS TREINADOS (.pkl — agora bundles model+scaler)"),
    (os.path.join(LOCAL_DIR, "reports"),                 "RELATÓRIOS (CSV, LaTeX, Markdown)"),
    (os.path.join(LOCAL_DIR, "explainability/results"),  "EXPLICABILIDADE (SHAP, Permutation, Ablação, Inovações)"),
    (os.path.join(LOCAL_DIR, "tuning/results"),          "TUNING (tabela de hiperparâmetros)"),
    (os.path.join(LOCAL_DIR, "utils/metadata"),          "METADADOS"),
]
PASTAS_DRIVE = [(DRIVE_DIR, "GOOGLE DRIVE — backup completo")]

print("FICHEIROS GERADOS PELO PIPELINE")
print("=" * 60)
total_colab = sum(listar_ficheiros(p, f"COLAB — {l}") for p, l in PASTAS_COLAB)
total_drive = sum(listar_ficheiros(p, l) for p, l in PASTAS_DRIVE)

print(f"\n{'='*60}")
print("  RESUMO FINAL")
print(f"{'='*60}")
print(f"  Ficheiros no Colab : {total_colab}")
print(f"  Ficheiros no Drive : {total_drive}")


FICHEIROS GERADOS PELO PIPELINE

  COLAB — DADOS (raw, clean, aggregated, features)
  /content/africa_mo_v4/data

  📁 aggregated/
     agregado_inner_join.csv  (328 KB)

  Total: 1 ficheiros  |  0.3 MB

  COLAB — MODELOS TREINADOS (.pkl — agora bundles model+scaler)
  /content/africa_mo_v4/models/artefacts

  📁 (raiz)/
     modelo_A1_WDI_only_Bayes_Complete.pkl  (19646 KB)
     modelo_A1_WDI_only_Bayes_Complete_MCMC.pkl  (19646 KB)
     modelo_A1_WDI_only_Bayes_Partial.pkl  (19708 KB)
     modelo_A1_WDI_only_Bayes_Partial_MCMC.pkl  (19708 KB)
     modelo_A1_WDI_only_LSTM.pkl  (169 KB)
     modelo_A1_WDI_only_RandomForest.pkl  (540 KB)
     modelo_A1_WDI_only_SARIMAX.pkl  (46 KB)
     modelo_A1_WDI_only_XGBoost.pkl  (143 KB)
     modelo_A2_WDI_PCA1_Bayes_Complete.pkl  (19646 KB)
     modelo_A2_WDI_PCA1_Bayes_Complete_MCMC.pkl  (19646 KB)
     modelo_A2_WDI_PCA1_Bayes_Partial.pkl  (19709 KB)
     modelo_A2_WDI_PCA1_Bayes_Partial_MCMC.pkl  (19709 KB)
     modelo_A2_WDI_PCA1_LSTM.pkl  (173

**Baixar todos os ficheiros gerados**

In [19]:
import os, zipfile, shutil

LOCAL_DIR = "/content/africa_mo_v4"
DRIVE_DIR = "/content/drive/MyDrive/africa_mo_pipeline"
ZIP_PATH  = "/content/pipeline_resultados_completos.zip"

PASTAS = [
    os.path.join(LOCAL_DIR, "data"),
    os.path.join(LOCAL_DIR, "models", "artefacts"),
    os.path.join(LOCAL_DIR, "reports"),
    os.path.join(LOCAL_DIR, "explainability", "results"),
    os.path.join(LOCAL_DIR, "tuning", "results"),
    os.path.join(LOCAL_DIR, "utils", "metadata"),
]

print("A criar ZIP com todos os resultados...")
n_ficheiros, tamanho = 0, 0
with zipfile.ZipFile(ZIP_PATH, "w", compression=zipfile.ZIP_DEFLATED) as zf:
    for pasta in PASTAS:
        if not os.path.exists(pasta):
            print(f"  ✗ Não existe: {pasta}")
            continue
        label = os.path.relpath(pasta, LOCAL_DIR)
        for dirpath, dirnames, filenames in os.walk(pasta):
            dirnames[:] = [d for d in dirnames if d != "__pycache__"]
            for f in sorted(filenames):
                if f.endswith(".pyc"):
                    continue
                caminho_abs = os.path.join(dirpath, f)
                caminho_zip = os.path.join("pipeline_resultados", label, os.path.relpath(caminho_abs, pasta))
                zf.write(caminho_abs, caminho_zip)
                n_ficheiros += 1
                tamanho += os.path.getsize(caminho_abs) / 1024

tamanho_zip = os.path.getsize(ZIP_PATH) / 1024 / 1024
print(f"\n✓ ZIP criado: {ZIP_PATH}")
print(f"  Ficheiros incluídos : {n_ficheiros}")
print(f"  Tamanho original    : {tamanho/1024:.1f} MB")
print(f"  Tamanho comprimido  : {tamanho_zip:.1f} MB")

shutil.copy(ZIP_PATH, os.path.join(DRIVE_DIR, "pipeline_resultados_completos.zip"))
print(f"\n✓ Cópia no Drive: {DRIVE_DIR}/pipeline_resultados_completos.zip")

from google.colab import files
files.download(ZIP_PATH)
print("✓ Download iniciado")


A criar ZIP com todos os resultados...

✓ ZIP criado: /content/pipeline_resultados_completos.zip
  Ficheiros incluídos : 170
  Tamanho original    : 400.5 MB
  Tamanho comprimido  : 374.1 MB

✓ Cópia no Drive: /content/drive/MyDrive/africa_mo_pipeline/pipeline_resultados_completos.zip


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

✓ Download iniciado
